diff --git a/drivers/mtd/devices/m25p80.c b/drivers/mtd/devices/m25p80.c
index 00eea6f..13756c7 100755
--- a/drivers/mtd/devices/m25p80.c
+++ b/drivers/mtd/devices/m25p80.c
@@ -56,6 +56,20 @@ static void m25p_addr2cmd(struct spi_nor *nor, unsigned int addr, u8 *cmd)
 	cmd[4] = addr >> (nor->addr_width * 8 - 32);
 }

+static void m25p_shutdown(struct spi_device *spi)
+{
+    struct m25p    *flash = spi_get_drvdata(spi);
+    
+    // manfeel note: add spi flash reset code
+    flash->command[0] = 0x66;
+    spi_write(flash->spi, flash->command, 1);
+    flash->command[0] = 0x99;
+    spi_write(flash->spi, flash->command, 1);
+    
+    /* Clean up MTD stuff. */
+    mtd_device_unregister(&flash->mtd);   
+}
+
 static int m25p_cmdsz(struct spi_nor *nor)
 {
 	return 1 + nor->addr_width;
@@ -385,6 +399,8 @@ static struct spi_driver m25p80_driver = {
 	.id_table	= m25p_ids,
 	.probe	= m25p_probe,
 	.remove	= m25p_remove,
+    // manfeel, add shutdown method to reset spi flash
+    .shutdown = m25p_shutdown,
 
 	/* REVISIT: many of these chips have deep power-down modes, which
 	 * should clearly be entered on suspend() to minimize power use.
